### PR TITLE
[RevEng] Output the HasColumnType() and HasMaxLength() APIs correctly

### PR DESCRIPTION
**Pull request by lajones**
Fix for #4312.

For SQL Server this means for `nvarchar` columns (or synonyms) we use `HasMaxLength()` and not `HasColumnType()`, for other string types we use `HasColumnType()` and not `HasMaxLength()`. Similarly for `varbinary` (or synonyms) use `HasMaxLength()` and not `HasColumnType()`, but for other binary column types use `HasColumnType()` and not `HasMaxLength()`.

### CODE REVIEW

File: src/EntityFramework.MicrosoftSqlServer.Design/SqlServerScaffoldingModelFactory.cs [hunk 1]
```diff
@@ -18,6 +19,11 @@ public class SqlServerScaffoldingModelFactory : RelationalScaffoldingModelFactor
     {
         private const int DefaultTimeTimePrecision = 7;
 
+        private static readonly ISet<string> _dataTypesForbiddingLength =
```
Comment (natemcmaster) [inline src/EntityFramework.MicrosoftSqlServer.Design/SqlServerScaffoldingModelFactory.cs]: Anyway we can capture this in the SqlServer type mapper? I thought it was already designed to handle lengths.

File: src/EntityFramework.MicrosoftSqlServer.Design/SqlServerScaffoldingModelFactory.cs [hunk 2]
```diff
@@ -144,16 +145,64 @@ private PropertyBuilder VisitTypeMapping(PropertyBuilder propertyBuilder, Column
...
+                    }
+                }
             }
 
             return propertyBuilder;
         }
 
+        private bool IsTypeAlias(ColumnModel column)
+        {
+            var typeAliases = column.Table.Database.SqlServer().TypeAliases;
```
Comment (natemcmaster) [inline src/EntityFramework.MicrosoftSqlServer.Design/SqlServerScaffoldingModelFactory.cs]: Possible to simplify?
`=> column.Table.Database.SqlServer().TypeAliases?.ContainsKey(column.DataType) == true;`

File: src/EntityFramework.MicrosoftSqlServer.Design/SqlServerScaffoldingModelFactory.cs [hunk 3]
```diff
@@ -144,16 +145,64 @@ private PropertyBuilder VisitTypeMapping(PropertyBuilder propertyBuilder, Column
...
+                        // what we did above).
+                        propertyBuilder.Metadata.SetMaxLength(null);
+                    }
+                }
             }
 
             return propertyBuilder;
         }
 
+        private bool IsTypeAlias(ColumnModel column)
```
Comment (natemcmaster) [inline src/EntityFramework.MicrosoftSqlServer.Design/SqlServerScaffoldingModelFactory.cs]: `HasTypeAlias`?

**Comment by natemcmaster** [comment]
Seems like we could make improvements to reuse type mapping API. Perhaps we need new API, such as `IRelationalTypeMapper.FindMapping(string typeName, int length)`. We already have some API for handling sized type mappings (`RelationalSizedTypeMapping`)

Thoughts on this @ajcvickers ?

File: src/EntityFramework.MicrosoftSqlServer.Design/SqlServerScaffoldingModelFactory.cs (hunk 3 above)
Comment (lajones) [inline src/EntityFramework.MicrosoftSqlServer.Design/SqlServerScaffoldingModelFactory.cs]: Sure.

File: src/EntityFramework.MicrosoftSqlServer.Design/SqlServerScaffoldingModelFactory.cs (hunk 2 above)
Comment (lajones) [inline src/EntityFramework.MicrosoftSqlServer.Design/SqlServerScaffoldingModelFactory.cs]: OK. Will do.

**Comment by lajones** [comment]
@natemcmaster @ajcvickers Nit: It would definitely have to be `int? length`.

The problem is that different data types interpret having a null max length differently. For `image`, `text` etc it is legal (in fact expected) and means "max length doesn't apply to this data type", for `nvarchar`, `varbinary` etc it means "equivalent to `nvarchar(max)`, `varbinary(max)`" etc. For other types it would make no sense e.g. `nchar`, `char` and would be an error.

**Comment by natemcmaster** [comment]
>  It would definitely have to be int? length.

Or an overload.

> The problem is that different data types interpret having a null max length differently....

To clarify, my point was that this logic could be placed in `SqlServerTypeMapper` instead of `SqlServerScaffoldModelFactory`.

**Comment by lajones** [comment]
@natemcmaster Sort of. This is not just about type mapping. It would be perfectly legal for us to output e.g. `HasColumnType("nvarchar(100)")` but we have chosen instead to write out `HasMaxLength(100)` and rely on the fact that the code we produce will define the CLR type for that column as a `string` and the default mapping for a string is back to `nvarchar`.

So in the end we need 2 pieces of information out: 1) should we use `HasColumnType()` or `HasMaxLength()` and 2) if `HasColumnType()`, what is the full definition of the column type we want to use, including any length qualifications.

I'll try to put all of that in `SqlServerTypeMapper` and see what it looks like.

**Comment by lajones** [comment]
@ajcvickers Have updated as suggested above - could you give the final sign-off since Nate is away?

**Comment by lajones** [comment]
Removing "Needs Review" while I rebase this on the renamed codebase and get it working again. Will put it back on when ready for review again.

File: src/EntityFramework.MicrosoftSqlServer/Storage/Internal/SqlServerTypeMapper.cs [hunk 4]
```diff
@@ -36,6 +38,11 @@ public class SqlServerTypeMapper : RelationalTypeMapper
         private readonly RelationalTypeMapping _decimal = new RelationalTypeMapping("decimal(18, 2)", typeof(decimal));
         private readonly RelationalTypeMapping _time = new RelationalTypeMapping("time", typeof(TimeSpan));
 
+        private static readonly ISet<string> _dataTypesForbiddingMaxLength =
+            new HashSet<string>(StringComparer.OrdinalIgnoreCase) { "image", "ntext", "text" };
```
Comment (jdlundquist) [inline src/EntityFramework.MicrosoftSqlServer/Storage/Internal/SqlServerTypeMapper.cs]: What about rowversion and uniqueidentifier - would these map to byte[] and string CLR types, respectively, and need to be excluded?

Also, I don't think date, datetime, or smalldatetime allow max length (at least per SQL Server, not sure about `HasMaxLength()`), but those would be excluded implicitly by CLR type in `MaxLengthQualifiedDataType`

File: src/EntityFramework.MicrosoftSqlServer/Storage/Internal/SqlServerTypeMapper.cs [hunk 5]
```diff
@@ -117,6 +124,66 @@ public SqlServerTypeMapper()
...
+        /// <param name="maxLength">
+        ///   the max length of the same column as defined by SQL Server metadata
+        /// </param>
+        /// <returns>
+        ///   a tuple of the max-length-qualified SQL Server type name, e.g. nvarchar(max),
+        ///   which can be used by HasColumnType() and the max length that should be
+        ///   defined on that same property through HasMaxLength()
+        /// </returns>
+        public virtual Tuple<string, int?> MaxLengthQualifiedDataType(
+            [NotNull] string unqualifiedTypeName, int? maxLength)
```
Comment (jdlundquist) [inline src/EntityFramework.MicrosoftSqlServer/Storage/Internal/SqlServerTypeMapper.cs]: What about decimal/numeric that take a precision and scale vs. a single max length? anything to consider for that or outside the scope of #4312 ?

File: src/EntityFramework.MicrosoftSqlServer/Storage/Internal/SqlServerTypeMapper.cs (hunk 4 above)
Comment (lajones) [inline src/EntityFramework.MicrosoftSqlServer/Storage/Internal/SqlServerTypeMapper.cs]: Please don't review just yet - I'm in the process of updating this and will add the "Needs Review" tag back when it's ready. Nonetheless I take your point and will check those scenarios.

**Comment by lajones** [comment]
I tried to put the code in `SqlServerTypeMapper` but what we are doing in scaffolding right there is overriding already assigned column types and max lengths. And sometimes it's important _not_ to override the existing values - so it's not really a generic "output column type and max length for this column" method which would be more generally useful throughout the stack - it's quite specific to what scaffolding needs. I agree such a method would be useful and I'll raise another issue to suggest looking into `SqlServerTypeMapper` and redesigning so that there is such a method. But for this issue, after discussion with @natemcmaster, I've moved most of the code back into `SqlServerScaffoldingModelFactory`. I did have to leave in one change to `SqlServerTypeMapper` as a `binary` data type should not be mapped to `_varbinary` otherwise it thinks its DefaultTypeName is `varbinary` and we end up outputting `HasMaxLength()` instead of `HasColumnType()`.

That said there are real synonyms for some types e.g. `varbinary` has the synonym `binary varying`. So I've updated the tests so we now test DDL using all data types including synonyms and check that we get the correct `HasColumnType()` and `HasMaxLength()` fluent API as a result of that.

**Comment by natemcmaster** [comment]
LGTM. :shipit:

**Comment by lajones** [comment]
Checked in with commit 443c52b.

**Comment by lajones** [comment]
Note: opened issue #4425 to track trying to make the `SqlServerTypeMapper` APIs more useful as discussed with @natemcmaster.

**Pull request merged**